portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,101 @@
+/** Servlet that handles data of comment section. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
-  }
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery results = datastore.prepare(query);
+
+        List<Comment> comments = new ArrayList<>();
+        for (Entity entity : results.asIterable()) {
+            long id = entity.getKey().getId();
+            String data = (String) entity.getProperty("data");
+            String userEmail = (String) entity.getProperty("userEmail");
+            long timestamp = (long) entity.getProperty("timestamp");
+
+            Comment comment = new Comment(id, data, userEmail, timestamp);
+            comments.add(comment);
+        }
+        
+        Gson gson = new Gson();
+
+        response.setContentType("application/json;");
+        response.getWriter().println(gson.toJson(comments));
+    }
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        UserService userService = UserServiceFactory.getUserService();
+
+        if (!userService.isUserLoggedIn()) {
+            response.sendRedirect("/index.html");
+            return;
+        }
+        
+        String data = request.getParameter("comment");
+        String userEmail = userService.getCurrentUser().getEmail();
+
+        if (checkValid(data)) {
+            long timestamp = System.currentTimeMillis();
+
+            Entity commentEntity = new Entity("Comment");
+            commentEntity.setProperty("data", data);
+            commentEntity.setProperty("userEmail", userEmail);
+            commentEntity.setProperty("timestamp", timestamp);
+
+            DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+            datastore.put(commentEntity);
+        }
+
+        response.sendRedirect("/index.html");
+    }
+
+    /** Checks if comment is valid or not. */
+    private boolean checkValid(String data) {
+        boolean valid = true;
+
+        if (data.length() > 0) {
+            boolean text_exist = false;
+            for (int i = 0; i < data.length(); ++i) {
+                char cur = data.charAt(i);
+                if (cur != ' ' && cur != '\n' && cur != '\t' && cur != '\r') {
+                    text_exist = true;
+                    break;
+                }
+            }
+
+            if (!text_exist) {
+                valid = false;
+            }
+        } else {
+            valid = false;
+        }
```

Comment:
I am guessing you're trying to validate if char at 0th index is not a new line/tab etc. this can be simplified a great deal + please explore using java 8 streams, which will make this concise are readable.

Reply:
This function checks if there is any character in comment other than newlines / tabs / whitespaces , because that won't be a valid comment. Do you have any suggestions for this?